# Capstone Project - car collisions 

## The Query Objective

The dataset of Seattle Car Collisions has data of pedestrians, cyclists and vehicles involved in collisions. The main objective using this dataset is to explore 
- how many collisions involve cyclists and pedestrians
- whether the driver was distracted (inattention) was a factor
- the frequency of cyclist collisions using date and time data
- what areas have the most frequency of cyclist collisions
<br>
<br>
The audience for this project is for pedestrians and cyclists in the Seattle area. 

In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import folium
import seaborn as sb

In [29]:
df = pd.read_csv('Collisions.csv')
df.head()

X          Y  OBJECTID  INCKEY  COLDETKEY REPORTNO   STATUS  \
0 -122.356511  47.517361         1  327920     329420  3856094  Matched   
1 -122.361405  47.702064         2   46200      46200  1791736  Matched   
2 -122.317414  47.664028         3    1212       1212  3507861  Matched   
3 -122.318234  47.619927         4  327909     329409  EA03026  Matched   
4 -122.351724  47.560306         5  104900     104900  2671936  Matched   

       ADDRTYPE   INTKEY                                           LOCATION  \
0  Intersection  34911.0                      17TH AVE SW AND SW ROXBURY ST   
1         Block      NaN     HOLMAN RD NW BETWEEN 4TH AVE NW AND 3RD AVE NW   
2         Block      NaN  ROOSEVELT WAY NE BETWEEN NE 47TH ST AND NE 50T...   
3  Intersection  29054.0                           11TH AVE E AND E JOHN ST   
4         Block      NaN  WEST MARGINAL WAY SW BETWEEN SW ALASKA ST AND ...   

   ... ROADCOND                LIGHTCOND PEDROWNOTGRNT SDOTCOLNUM SPEEDING  \
0  ...      Dry                 Daylight           NaN        NaN      NaN   
1  ...      Wet                     Dusk           NaN  5101020.0      NaN   
2  ...      Dry  Dark - Street Lights On           NaN        NaN      NaN   
3  ...      Wet  Dark - Street Lights On           NaN        NaN      NaN   
4  ...      Ice  Dark - Street Lights On           NaN  9359012.0        Y   

   ST_COLCODE                                         ST_COLDESC  SEGLANEKEY  \
0          10                                  Entering at angle           0   
1          13  From same direction - both going straight - bo...           0   
2          30               From opposite direction - all others           0   
3           0             Vehicle going straight hits pedestrian           0   
4          50                                       Fixed object           0   

   CROSSWALKKEY  HITPARKEDCAR  
0             0             N  
1             0             N  
2             0             N  
3             0             N  
4             0             N  

[5 rows x 40 columns]

In [9]:
df.columns

Index(['X', 'Y', 'OBJECTID', 'INCKEY', 'COLDETKEY', 'REPORTNO', 'STATUS',
       'ADDRTYPE', 'INTKEY', 'LOCATION', 'EXCEPTRSNCODE', 'EXCEPTRSNDESC',
       'SEVERITYCODE', 'SEVERITYDESC', 'COLLISIONTYPE', 'PERSONCOUNT',
       'PEDCOUNT', 'PEDCYLCOUNT', 'VEHCOUNT', 'INJURIES', 'SERIOUSINJURIES',
       'FATALITIES', 'INCDATE', 'INCDTTM', 'JUNCTIONTYPE', 'SDOT_COLCODE',
       'SDOT_COLDESC', 'INATTENTIONIND', 'UNDERINFL', 'WEATHER', 'ROADCOND',
       'LIGHTCOND', 'PEDROWNOTGRNT', 'SDOTCOLNUM', 'SPEEDING', 'ST_COLCODE',
       'ST_COLDESC', 'SEGLANEKEY', 'CROSSWALKKEY', 'HITPARKEDCAR'],
      dtype='object')

In [14]:
df.shape

(221738, 40)

## Background Information

This dataset is of all the collisions provided by Seattle Police Department (SPD) and recorded by Traffic Records.  This dataset includes all types of collisions and datatypes. The timeframe is from 2004 to Oct 1 2020  when the dataset was downloaded. 

## Data Description

The dataset has 40 columns and over 221,000 rows of data. The columns that will be used for the Data Understanding will be related to bicyles, injuries, number of vehicles involved in collisions, date and time of collisions, and the factor of the driver paying attention.

<h4> columns of interest description </h4>

- PERSONCOUNT = total number of people involved in the collision
- PEDCOUNT = number of pedestrians involved in the collision. This is entered by the state.
- PEDCYLCOUNT = number of bicycles involved in the collision. This is entered by the state
- VEHCOUNT = number of vehicles involved in the collision. This is entered by the state.
- INJURIES = number of total injuries in the collision. This is entered by the state.
- SERIOUSINJURIES = number of serious injuries in the collision. This is entered by the state.
- FATALITIES =  number of fatalities in the collision. This is entered by the state.

- INCDTTM = The date and time of the incident.
- INATTENTIONIND = Whether or not collision was due to inattention. (Y/N)

- LOCATION = Description of the general location of the collision
- ADDRTYPE = Collision address type: Alley, Block, Intersection

## Data Preparation

In [16]:
# check for missing values in df
missing_values = df.isnull().sum()
missing_values

X                    7478
Y                    7478
OBJECTID                0
INCKEY                  0
COLDETKEY               0
REPORTNO                0
STATUS                  0
ADDRTYPE             3714
INTKEY             149711
LOCATION             4593
EXCEPTRSNCODE      120403
EXCEPTRSNDESC      209953
SEVERITYCODE            1
SEVERITYDESC            0
COLLISIONTYPE       26451
PERSONCOUNT             0
PEDCOUNT                0
PEDCYLCOUNT             0
VEHCOUNT                0
INJURIES                0
SERIOUSINJURIES         0
FATALITIES              0
INCDATE                 0
INCDTTM                 0
JUNCTIONTYPE        11979
SDOT_COLCODE            1
SDOT_COLDESC            1
INATTENTIONIND     191550
UNDERINFL           26431
WEATHER             26641
ROADCOND            26560
LIGHTCOND           26730
PEDROWNOTGRNT      216543
SDOTCOLNUM          94533
SPEEDING           211802
ST_COLCODE           9413
ST_COLDESC          26451
SEGLANEKEY              0
CROSSWALKKEY

<h4> columns of interest with any missing values summarized </h4>

   - PERSONCOUNT             0
   - PEDCOUNT                0
   - PEDCYLCOUNT             0
   - VEHCOUNT                0
   - INJURIES                0
   - SERIOUSINJURIES         0
   - FATALITIES              0
   - INCDATE                 0
   - INCDTTM                 0
   - INATTENTIONIND     191550
   - LOCATION             4593
   - ADDRTYPE             3714


column INATTENTIONIND has many missing values

In [24]:
# find the the number of misisng values for just column LOCATION
location_missing = df['LOCATION'].isnull().sum()
location_missing

4593

In [25]:
# get the total missing values from dataframe
total_missing = np.product(df.shape)
total_missing 

8869520

In [26]:
TOTAL_missing = location_missing.sum()
TOTAL_missing 

4593

In [27]:
# determine the percent of LOCATION data with missing values represents 
percent_missing = (TOTAL_missing / total_missing  ) *100
percent_missing 

0.05178408752672072

The total of all LOCATION missing values of dataset represents only 0.05% of data. This column can be used later for map visualizations

In [30]:
df_dropped = df.dropna(axis=1)
df_dropped.head()

OBJECTID  INCKEY  COLDETKEY REPORTNO   STATUS  \
0         1  327920     329420  3856094  Matched   
1         2   46200      46200  1791736  Matched   
2         3    1212       1212  3507861  Matched   
3         4  327909     329409  EA03026  Matched   
4         5  104900     104900  2671936  Matched   

                     SEVERITYDESC  PERSONCOUNT  PEDCOUNT  PEDCYLCOUNT  \
0  Property Damage Only Collision            2         0            0   
1  Property Damage Only Collision            2         0            0   
2                Injury Collision            2         0            0   
3                Injury Collision            3         1            0   
4                Injury Collision            2         0            0   

   VEHCOUNT  INJURIES  SERIOUSINJURIES  FATALITIES                 INCDATE  \
0         2         0                0           0  2020/01/19 00:00:00+00   
1         2         0                0           0  2005/04/11 00:00:00+00   
2         2         1                0           0  2013/03/31 00:00:00+00   
3         1         1                0           0  2020/01/06 00:00:00+00   
4         1         1                0           0  2009/12/25 00:00:00+00   

                 INCDTTM  SEGLANEKEY  CROSSWALKKEY HITPARKEDCAR  
0   1/19/2020 9:01:00 AM           0             0            N  
1   4/11/2005 6:31:00 PM           0             0            N  
2   3/31/2013 2:09:00 AM           0             0            N  
3    1/6/2020 5:55:00 PM           0             0            N  
4  12/25/2009 7:00:00 PM           0             0            N

In [18]:
df_dropped.shape

(221738, 18)

In [19]:
df_dropped.isnull().sum()

OBJECTID           0
INCKEY             0
COLDETKEY          0
REPORTNO           0
STATUS             0
SEVERITYDESC       0
PERSONCOUNT        0
PEDCOUNT           0
PEDCYLCOUNT        0
VEHCOUNT           0
INJURIES           0
SERIOUSINJURIES    0
FATALITIES         0
INCDATE            0
INCDTTM            0
SEGLANEKEY         0
CROSSWALKKEY       0
HITPARKEDCAR       0
dtype: int64

The df_dropped dataset now has 18 columns wide instead of 40 with no missing values. The columns of interest are all numeric and ready for use. 